internal/datastore/crdb/watch.go

```diff
@@ -193,7 +202,130 @@ func (cds *crdbDatastore) watch(
+	// we do not accumulate in this implementation, but stream right away
+	return nil, nil
+}
+
+func (s StreamingChangeProvider) AddRelationshipChange(ctx context.Context, rev revisions.HLCRevision, rel tuple.Relationship, op tuple.UpdateOperation) error {
+	if s.content&datastore.WatchRelationships != datastore.WatchRelationships {
+		return nil
+	}
+
+	changes := datastore.RevisionChanges{
+		Revision: rev,
+	}
+	switch op {
+	case tuple.UpdateOperationCreate:
+		changes.RelationshipChanges = append(changes.RelationshipChanges, tuple.Create(rel))
+	case tuple.UpdateOperationTouch:
+		changes.RelationshipChanges = append(changes.RelationshipChanges, tuple.Touch(rel))
+	case tuple.UpdateOperationDelete:
+		changes.RelationshipChanges = append(changes.RelationshipChanges, tuple.Delete(rel))
+	default:
+		return spiceerrors.MustBugf("unknown change operation")
+	}
+
+	return s.sendChange(&changes)
+}
+
+func (s StreamingChangeProvider) AddChangedDefinition(_ context.Context, rev revisions.HLCRevision, def datastore.SchemaDefinition) error {
+	if s.content&datastore.WatchSchema != datastore.WatchSchema {
+		return nil
+	}
+
+	changes := datastore.RevisionChanges{
+		Revision:           rev,
+		ChangedDefinitions: []datastore.SchemaDefinition{def},
+	}
+
+	return s.sendChange(&changes)
+}
+
+func (s StreamingChangeProvider) AddDeletedNamespace(_ context.Context, rev revisions.HLCRevision, namespaceName string) error {
+	if s.content&datastore.WatchSchema != datastore.WatchSchema {
+		return nil
+	}
+
+	changes := datastore.RevisionChanges{
+		Revision:          rev,
+		DeletedNamespaces: []string{namespaceName},
+	}
+
+	return s.sendChange(&changes)
+}
+
+func (s StreamingChangeProvider) AddDeletedCaveat(_ context.Context, rev revisions.HLCRevision, caveatName string) error {
+	if s.content&datastore.WatchSchema != datastore.WatchSchema {
+		return nil
+	}
+
+	changes := datastore.RevisionChanges{
+		Revision:       rev,
+		DeletedCaveats: []string{caveatName},
+	}
+
+	return s.sendChange(&changes)
+}
+
+func (s StreamingChangeProvider) SetRevisionMetadata(_ context.Context, rev revisions.HLCRevision, metadata map[string]any) error {
+	if len(metadata) > 0 {
+		parsedMetadata, err := structpb.NewStruct(metadata)
+		if err != nil {
+			return spiceerrors.MustBugf("failed to convert metadata to structpb: %v", err)
+		}
+
+		changes := datastore.RevisionChanges{
+			Revision: rev,
+			Metadata: parsedMetadata,
+		}
+
+		return s.sendChange(&changes)
+	}
+
+	return nil
+}
+
+type (
+	sendChangeFunc func(change *datastore.RevisionChanges) error
+	sendErrorFunc  func(err error)
+)
+
+func (cds *crdbDatastore) processChanges(ctx context.Context, changes pgx.Rows, sendError sendErrorFunc, sendChange sendChangeFunc, opts datastore.WatchOptions, streaming bool) {
+	var tracked ChangeTracker[revisions.HLCRevision, revisions.HLCRevision]
+	if streaming {
+		tracked = &StreamingChangeProvider{
```

Comment:
why is this exported?

Reply:
it does not have to, this is the internal package. I'm ok making it unexported, and same for the interface

Reply:
Please do

Reply:
done